### ISSUE TITLE
uwsgi errors on docker deployment

### ISSUE THREAD

**Issue by EnzoAkira** [alerta/alerta#1168]
**Issue Summary**
- Alerta Webui does not open.
- I get housekeeping and heartbeat errors like: `Failed to get config from http://localhost:8080/api/config: Reason: not a JSON object`

**Environment**
- OS: Linux
- API version: 7.4.4
- Deployment: Docker (docker-compose)

- Database: Postgres on main host

- Server config:
> alertad.conf
```ini
   SIGNUP_ENABLED = False
   DEFAULT_NORMAL_SEVERITY = 'ok'
   SEVERITY_MAP = {
     'critical': 1,
     'warning': 4,
     'ok': 5,
     'cleared': 9,
     'unknown': 9,
     'indeterminate': 8
    }
   COLOR_MAP = {
     'severity': {
       'critical': '#c02d3b',
       'warning': '#fe7d37',
       'ok': '#3e961d',
       'cleared': '#8bc077',
       'unknown': '#9f9f9f',
       'indeterminate': 'silver'
     },
    'text': 'black',
    'highlight': '#87CEEB'
   }
   CORS_ORIGINS = [
      'http://localhost:8080/',
      'http://localhost:9900/',
      r'http://\w*:9900/',
     r'http?://\w*\.?my\-server\.com:?\d*/?.*',
    r'http?://\w*\.?myserver\.es:?\d*/?.*'
   ]
   ALLOWED_ENVIRONMENTS = ['prod', 'dev']
   JSON_AS_ASCII=False
   JSON_SORT_KEYS=True
   JSONIFY_PRETTYPRINT_REGULAR=True
``` 
---

- web UI version: 7.4.4
- CLI version: 7.4.0

**To Reproduce**
1. Deploy with docker-compose
```yml
version: '3.1'
services:
  web:
    image: alerta/alerta-web
    container_name: alerta-server
    volumes:
      - $PWD/alertad.conf:/app/alertad.conf
    ports:
      - 9900:8080
    environment:
      - DEBUG=1  # remove this line to turn DEBUG off
      - AUTH_REQUIRED=True
      - PLUGINS=reject,heartbeat,blackout
      - DATABASE_URL=postgres://alerta:Password@172.17.0.1/alerta_db
    restart: always
```
3. Access to http://myserver:9900
4. Get error on config.json
5. On container logs:
```
2020-03-11 13:06:46,229 DEBG 'nginx' stdout output:
ip=\- [\11/Mar/2020:13:06:46 +0000] "\GET /api/config HTTP/1.1" \200 \994 "\-" "\python-requests/2.21.0"
/web/ | /index.html | > GET /api/config HTTP/1.1

2020-03-11 13:06:46,230 DEBG 'housekeeping' stdout output:
Error: Failed to get config from http://localhost:8080/api/config: Reason: not a JSON object

2020-03-11 13:06:46,232 DEBG 'nginx' stdout output:
2020/03/11 13:06:46 [info] 22#22: *75 client 127.0.0.1 closed keepalive connection

2020-03-11 13:06:46,264 DEBG fd 8 closed, stopped monitoring <POutputDispatcher at 140647628668224 for <Subprocess at 140647628788368 with name housekeeping in state RUNNING> (stdout)>
2020-03-11 13:06:46,265 INFO exited: housekeeping (exit status 1; not expected)
2020-03-11 13:06:46,265 DEBG received SIGCLD indicating a child quit
2020-03-11 13:06:47,267 INFO spawned: 'housekeeping' with pid 227
2020-03-11 13:06:47,309 DEBG 'nginx' stdout output:
ip=\- [\11/Mar/2020:13:06:47 +0000] "\GET /api/config HTTP/1.1" \200 \994 "\-" "\python-requests/2.21.0"
/web/ | /index.html | > GET /api/config HTTP/1.1

2020-03-11 13:06:47,311 DEBG 'nginx' stdout output:
2020/03/11 13:06:47 [info] 22#22: *76 client 127.0.0.1 closed keepalive connection

2020-03-11 13:06:47,311 DEBG 'heartbeats' stdout output:
Error: Failed to get config from http://localhost:8080/api/config: Reason: not a JSON object

2020-03-11 13:06:47,338 DEBG fd 6 closed, stopped monitoring <POutputDispatcher at 140647628805976 for <Subprocess at 140647628788872 with name heartbeats in state RUNNING> (stdout)>
2020-03-11 13:06:47,339 INFO exited: heartbeats (exit status 1; not expected)
2020-03-11 13:06:47,339 DEBG received SIGCLD indicating a child quit
2020-03-11 13:06:48,342 INFO spawned: 'heartbeats' with pid 229
2020-03-11 13:06:48,342 INFO success: housekeeping entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2020-03-11 13:06:49,343 INFO success: heartbeats entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
```

**Comment by satterly** [comment]
Remove the leading spaces from your configuration file.

**Comment by EnzoAkira** [comment]
Sorry, the double spaces were only here. My configuration file is fine.
Do you need more info to resolve the issue?

**Comment by satterly** [comment]
What is the response you get from `http://localhost:9900/api/config`? It should be JSON config.

**Comment by EnzoAkira** [comment]
- If I run curl on main host to `http://localhost:9900/api/config` I get the index.html

<details>
  <summary>curl response</summary>
    <!DOCTYPE html><html lang=en><head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width,initial-scale=1"><title>Alerta</title><meta name=application-name content=Alerta><link rel=apple-touch-icon-precomposed sizes=57x57 href=/apple-touch-icon-57x57.png><link rel=apple-touch-icon-precomposed sizes=114x114 href=/apple-touch-icon-114x114.png><link rel=apple-touch-icon-precomposed sizes=72x72 href=/apple-touch-icon-72x72.png><link rel=apple-touch-icon-precomposed sizes=144x144 href=/apple-touch-icon-144x144.png><link rel=apple-touch-icon-precomposed sizes=60x60 href=/apple-touch-icon-60x60.png><link rel=apple-touch-icon-precomposed sizes=120x120 href=/apple-touch-icon-120x120.png><link rel=apple-touch-icon-precomposed sizes=76x76 href=/apple-touch-icon-76x76.png><link rel=apple-touch-icon-precomposed sizes=152x152 href=/apple-touch-icon-152x152.png><link rel=icon type=image/png href=/favicon-196x196.png sizes=196x196><link rel=icon type=image/png href=/favicon-96x96.png sizes=96x96><link rel=icon type=image/png href=/favicon-32x32.png sizes=32x32><link rel=icon type=image/png href=/favicon-16x16.png sizes=16x16><link rel=icon type=image/png href=/favicon-128.png sizes=128x128><meta name=msapplication-TileColor content=#FFFFFF><meta name=msapplication-TileImage content=/mstile-144x144.png><meta name=msapplication-square70x70logo content=/mstile-70x70.png><meta name=msapplication-square150x150logo content=/mstile-150x150.png><meta name=msapplication-wide310x150logo content=/mstile-310x150.png><meta name=msapplication-square310x310logo content=/mstile-310x310.png><link rel=preload href=/fonts/Roboto-Regular.ttf as=font crossorigin><link rel=preload href=/fonts/MaterialIcons.woff2 as=font type=font/woff2 crossorigin><link rel=preload href=/fonts/Sintony-Regular.ttf as=font crossorigin><link rel=preload href=/fonts/Sintony-Bold.ttf as=font crossorigin><link href=/css/admin.026d6aa5.css rel=prefetch><link href=/css/user.d698dc05.css rel=prefetch><link href=/js/admin.480587ce.js rel=prefetch><link href=/js/auth.84f77df9.js rel=prefetch><link href=/js/user.e4ec56a1.js rel=prefetch><link href=/css/app.9f54de09.css rel=preload as=style><link href=/css/chunk-vendors.ab396d9c.css rel=preload as=style><link href=/js/app.cd253c4b.js rel=preload as=script><link href=/js/chunk-vendors.6a3b5dfa.js rel=preload as=script><link href=/css/chunk-vendors.ab396d9c.css rel=stylesheet><link href=/css/app.9f54de09.css rel=stylesheet></head><body><noscript><strong>We're sorry but Vue doesn't work properly without JavaScript enabled. Please enable it to continue.</strong></noscript><script>(function(){
        var redirect = sessionStorage.redirect;
        delete sessionStorage.redirect;
        if (redirect && redirect != location.href) {
            history.replaceState(null, null, redirect);
        }
      })();</script><div id=app></div><script src=/js/chunk-vendors.6a3b5dfa.js></script><script src=/js/app.cd253c4b.js></script></body></html>
</details>

- On browser with public address I get:

![imagem](https://user-images.githubusercontent.com/39481527/76474066-5951f400-63f2-11ea-81a5-00c431814683.png)
**And**
![imagem](https://user-images.githubusercontent.com/39481527/76474258-f745be80-63f2-11ea-9b62-35b7090c3547.png)

**Comment by guohaolee** [comment]
Hi, i seems to have this issue too. After investigating for some time, i noticed that its caused by the nginx.conf file in the docker container. 

The issue is at here: https://github.com/alerta/docker-alerta/blob/master/nginx.conf#L49
What i feel like it is not getting the  **!BASE_PATH!** correctly. That's why it is returning the error. Replacing the **!BASE_PATH!** with **/api** works fine for me. 

On the other note, I also notice that the **!WEB_PATH!** is not populated correctly as well at here: https://github.com/alerta/docker-alerta/blob/master/nginx.conf#L41

@satterly , can you advice on the right place to populate the **!BASE_PATH!** and **!WEB_PATH!**?

**Comment by EnzoAkira** [comment]
Hi everyone!.
In the past couple of weeks, I have been testing some modifications I made to Docker Alerta. I inform that in my case I am building the Alerta image locally.

**TL; DR**:
- Modify nginx.conf L# 33 to:
`listen *: 8080 default_server;`
- Add this to Dockerfile after L# 67:
`chmod -R u + s / app / venv / web && \`
- Edit your docker-compose.alerta.yml add:
```yml
...
web
  build:
   context:.
   dockerfile: Dockerfile
    args:
     VERSION: 7.4.5
  image: alert / web-alert: 7.4.5
```
- Run docker-compose

---

1. Modifying nginx.conf
After reading the nginx docs and several attempts, this was the solution to not give `Failed to get config` errors

2. Modifying the Dockerfile
In Dockerfile I had to add a new option. As @guohaolee points out, the replacement was not completely done by the `docker-entrypoint.sh` file. As far as I could tell, the script ran the 1st `sed` but the second gave an error in writing the file.
In this way all files can be written without problems, regardless of the user.
@satterly If you eventually have a better solution say it here!

My docker compose:
<details>
  <summary>docker-compose.alerta.yml</summary>
<p>

```yml
version: '3.1'
   services:
    web:
     build:
       context: .
       dockerfile: Dockerfile
       args:
         VERSION: 7.4.5
     image: alerta/alerta-web:7.4.5
     container_name: alerta-server
     volumes:
       - $PWD/config/alertad.conf:/app/alertad.conf
       - $PWD/config/alerta.conf:/app/alerta.conf
       - $PWD/config/config.json:/web/config.json
     ports:
       - "18000:8080"
     environment:
       - DEBUG=1  # remove this line to turn DEBUG off
       - AUTH_REQUIRED=True
     restart: on-failure
```

</p>
</details>

My Dockerfile:
<details>
  <summary>Dockerfile</summary>
<p>

```yml
FROM python:3.7.7
ENV PYTHONUNBUFFERED 1

LABEL maintainer="Nick Satterly <nick.satterly@gmail.com>"

ARG BUILD_DATE=now
ARG VCS_REF
ARG VERSION

ARG SERVER_VERSION=${VERSION}
ARG CLIENT_VERSION=7.4.4
ARG WEBUI_VERSION=${VERSION}

LABEL org.label-schema.build-date=$BUILD_DATE \
      org.label-schema.url="https://alerta.io" \
      org.label-schema.vcs-url="https://github.com/alerta/docker-alerta" \
      org.label-schema.vcs-ref=$VCS_REF \
      org.label-schema.version=$VERSION \
      org.label-schema.schema-version="1.0.0-rc.1"

SHELL ["/bin/bash", "-o", "pipefail", "-c"]

# hadolint ignore=DL3008
RUN curl -fsSL https://www.mongodb.org/static/pgp/server-4.2.asc | apt-key add - && \
    echo "deb https://repo.mongodb.org/apt/debian buster/mongodb-org/4.2 main" | tee /etc/apt/sources.list.d/mongodb-org-4.2.list && \
    apt-get update && \
    apt-get install -y --no-install-recommends \
    gettext-base \
    libffi-dev \
    libldap2-dev \
    libpq-dev \
    libsasl2-dev \
    nginx-light \
    postgresql-client \
    python3-dev \
    supervisor \
    wget \
    mongodb-org-shell \
    vim && \
    apt-get -y clean && \
    apt-get -y autoremove && \
    rm -rf /var/lib/apt/lists/*

COPY requirements.txt /app/requirements.txt
RUN pip install --no-cache-dir pip==20.0.2 virtualenv==20.0.1 && \
    python3 -m venv /venv && \
    /venv/bin/pip install --upgrade pip && \
    /venv/bin/pip install --upgrade setuptools==45.2.0 && \
    /venv/bin/pip install -r /app/requirements.txt
ENV PATH $PATH:/venv/bin

RUN /venv/bin/pip install alerta==${CLIENT_VERSION} alerta-server==${SERVER_VERSION}
COPY install-plugins.sh /app/install-plugins.sh
COPY plugins.txt /app/plugins.txt
RUN /app/install-plugins.sh

ADD https://github.com/alerta/alerta-webui/releases/download/v${WEBUI_VERSION}/alerta-webui.tar.gz /tmp/webui.tar.gz
RUN tar zxvf /tmp/webui.tar.gz -C /tmp && \
    mv /tmp/dist /web && \
    rm -f /tmp/webui.tar.gz
COPY config.json.template /web/config.json.template

COPY wsgi.py /app/wsgi.py
COPY uwsgi.ini /app/uwsgi.ini
COPY nginx.conf /app/nginx.conf

RUN ln -sf /dev/stdout /var/log/nginx/access.log
RUN ln -sf /dev/stdout /var/log/nginx/error.log
RUN chgrp -R 0 /app /venv /web && \
    chmod -R g=u /app /venv /web && \
    chmod -R u+s /app /venv /web && \
    useradd -u 1001 -g 0 alerta

USER 1001

ENV ALERTA_SVR_CONF_FILE=/app/alertad.conf
ENV ALERTA_CONF_FILE=/app/alerta.conf
ENV ALERTA_WEB_CONF_FILE=/web/config.json

ENV BASE_URL=/api
ENV HEARTBEAT_SEVERITY=critical

COPY docker-entrypoint.sh /usr/local/bin/
COPY supervisord.conf /app/supervisord.conf

ENTRYPOINT ["docker-entrypoint.sh"]

EXPOSE 8080
CMD ["supervisord", "-c", "/app/supervisord.conf"]
```

</p>
</details>

---

I also mention here that I have successfully tested a new plugin for Alerta, it is now possible to send and receive calls and sms to and from Messagebird if a user belongs to a specific group. For now I only have python done but soon I will deal with the instructions and respective documentation.
